### ISSUE TITLE
Remove some Duktape.xxx bindings

### ISSUE THREAD

**Issue by svaarala** [svaarala/duktape#866]
Some of the Duktape.xxx bindings are a sandboxing issue and could be removed or replaced with a C API (which would then allow an Ecmascript binding to be built if necessary). Go through the Duktape.xxx bindings for Duktape 2.x:

- [ ] Duktape.Buffer: to be removed as part of #865
- [ ] Duktape.Pointer
- [ ] Duktape.Thread
- [ ] Duktape.info: used almost exclusively for debugging / diagnostic, e.g. to check whether an integer is a fastint or how much memory an object/array takes; could be removed or replaced with a C API
- [ ] Duktape.act: useful but potential sandboxing issue because it allows access to function objects in the current call stack
- [ ] Duktape.gc: useful but probably enough to have a C API (`duk_gc()`)
- [ ] Duktape.fin
- [ ] Duktape.enc
- [ ] Duktape.dec
- [ ] Duktape.compact

**Comment by fatcerberus** [comment]
Isn't `Duktape.info()` used in some of the unit tests?  The fastint tests come to mind (testing whether fastint promotion/demotion works properly).

**Comment by fatcerberus** [comment]
`Duktape.fin()` should probably stay.  There's no other way to assign a finalizer in Ecmascript code, and forcing the app to provide an entry point is awkward since it's a pretty engine-specific thing.

Same for enc/dec, as they are needed to use JX/JC.

`Duktape.act()` I wouldn't miss as long as an equivalent C API is added.

**Comment by svaarala** [comment]
Regarding testcases, fastint tests indeed rely on Duktape.info() and would need to be removed or perhaps some C API could replace the Duktape.info() based fastint status detection. (It'd also be possible to expose a function for the unit tests without making it an official, versioned API call.)

I'm not sure I understand your argument re: Duktape.fin() - why would it be difficult/awkward to provide an Ecmascript binding which was implemented in terms of `duk_set_finalizer()`? I mean, why is this any more awkward than wrapping any other Duktape specific feature only available in C? Do you mean that every code base needing Ecmascript-driven finalization would need to do the same thing so that they wouldn't interoperate?

For finalizers there's also the current situation that Ecmascript finalizer callbacks are not fully robust - they don't work properly when a script timeout has been triggered. Because robust finalizers must be written in C, perhaps the whole API should be C only. There are also some sandboxing concerns w.r.t. finalizers; specifically, the thread running the finalizer may not be from the same global object space as where the object has been created, which may be a serious sandboxing concern in some cases.

**Comment by svaarala** [comment]
Added a related issue to the description: keeping all the Duktape specific stuff in the Duktape object is consistent in some sense, but it might be more natural for some of the custom functionality to reside as additional functions/properties of standard objects. A few developers have asked about this earlier on but 1.x was already released at that point.

For example, if the finalizer API is kept, maybe `Object.setFinalizer()` and `Object.getFinalizer()` would be more natural than `Duktape.fin()`. Similarly, `Duktape.enc(encoding, value)` might be more natural as `String(value, encoding)` (for example: `String({foo: 123}, 'jx')`). Or maybe `Object.encode()` and `Object.decode()` so that one could do `val.encode('jx')` for any value inheriting from Object.

This naturally carries some risk of conflict with standard properties, but is nevertheless an approach used by other engines for at least some custom features. The upside of doing so is that it may be possible to polyfill a compatible implementation in another engine (e.g. adding an encoding field to `String()`) without adding an actual Duktape object.

The presence of a `Duktape` object is also used for engine identification (and `Duktape.version` for engine version identification) but there are other approaches to that too; for example, a global `__duktapeVersion__` property or perhaps `Object.duktapeVersion` property. So, getting rid of the Duktape object entirely would also be feasible.

**Comment by fatcerberus** [comment]
Does the Ecmascript standard allow adding properties to `Object`?  I'm almost positive `Object.prototype` is off limits, but I'm not sure about `Object`.

**Comment by svaarala** [comment]
As far as I can remember the specification doesn't forbid additional properties? It just specifies what must be present. Almost all engines add custom stuff to `Object.prototype`; for example, V8 adds `Object.prototype.__defineGetter__` and a few others. Same for `Object`. It's of course a good idea to avoid property names that might conflict with future versions.

**Comment by svaarala** [comment]
So to be clear, I'm not sure which approach is better re: custom properties in Duktape object or elsewhere. If any changes are to be made, now's a good time to do that :)

I originally opted for using the Duktape object and to keep the standard objects as purely standard as possible. This has the upside that if you won't accidentally depend on a custom feature when writing code; any custom features you use will necessarily reference the `Duktape` object so it's apparent what's going on.

The upside of adding custom properties to standard objects is that it's more natural, it's used by other engines too, and adding fill-ins when moving from engine to another is more natural than writing a dummy `Duktape` stub on the other engine.

So, not an obvious choice by any means.

**Comment by fatcerberus** [comment]
I was under the impression it was forbidden by spec to add custom properties/methods to `Object.prototype` but I checked the E5.1 spec again, and that doesn't appear to actually be the case.

I'm not necessarily against moving some of the functions over to `Object`.  I think it would be cleaner, as you said--`Object.compact` being a great example.  However, this does have the downside that calling those functions isn't self-documenting w.r.t. where exactly the feature are supported.  Someone unfamiliar with the Ecmascript specification (which is quite large and complex) might think that, say, `Object.compact()` is standard and attempt to use the code with a different engine--where it will mysteriously fail (speaking from experience, having learned JavaScript at a time when SpiderMonkey included a lot of nonstandard stuff).  `Duktape.compact` makes it explicit that the feature is ONLY supported in Duktape.

I'm also quite partial to the elegance of `Duktape.version`.  The alternatives given for that seem much more awkward. :)

**Comment by fatcerberus** [comment]
Regarding `Duktape.fin()`: I was originally thinking of the case where an application provides an API call to set a finalizer implemented in terms of `duk_set_finalizer()` and then later that application wants to switch JS engines without breaking semantic versioning guarantees.  That becomes an issue if the new JS engine doesn't support, e.g. Ecmascript functions as finalizers.

However it now occurs to me that this scenario is problematic anyway--the non-Duktape engine is not likely to have `Duktape.fin()` after all.  So feel free to ignore that bit of stupidity on my part :)

**Comment by saghul** [comment]
> Duktape.fin() should probably stay. There's no other way to assign a finalizer in Ecmascript code, and forcing the app to provide an entry point is awkward since it's a pretty engine-specific thing.
> 

I share the feeling. I use it for closing C fds if my JS Socket objects are destroyed while active. Now I'd need to expose a wrapper for duk_set_finalizer I guess, but since it's a Duktape specific thing why not have it in the Duktape object?

> Same for enc/dec, as they are needed to use JX/JC.

Same as above. FWIW I'm currently using JX encoding for formatting output in a REPL (which I copied from the Duktape cmd example IIRC).

**Comment by svaarala** [comment]
@fatcerberus:

> However, this does have the downside that calling those functions isn't self-documenting w.r.t. where exactly the feature are supported

Right, and it's an actual issue when people move from other engines to Duktape, they're often surprised that some bindings like `Object.prototype.__defineGetter__` (which they thought to be standard) are missing. But even so, it seems to be the prevailing way to provide extensions at least when the extension can be clearly associated with a specific object type (`Object.xxx` for custom object operations like finalizers and compaction, etc).

**Comment by svaarala** [comment]
@saghul At a high level, Duktape doesn't provide (or attempt to provide) every operation in both C and Ecmascript versions. For example, there are forced property operations that are only possible from C code (which is important to avoid side-stepping Ecmascript property protections) and also operations that are also available in Ecmascript because mirroring them all in C would expand the API quite a bit.

So for every operation, the question isn't "why not both?" but "which versions (C / Ecmascript) make most sense" given how they're used, and also keeping in mind that API footprint has both a concrete cost (footprint) and various other costs like documenting, maintaining, and for the user to remember what's available (especially confusing if the options are not 100% equivalent).

As a concrete example, if I added *every* Ecmascript built-in operation to the C API, I would think the result would be worse than current because the size of the C API would probably roughly double. Learning the API would be more difficult, and there'd be natural user questions like "is the C version more efficient than the Ecmascript one; if so, by how much?", etc :)

That's why I'm trying to minimize overlap between the C and the Ecmascript APIs, and only keep overlap when there's a good use case for doing so. For some operations the answer is obvious (e.g. dealing with the value stack, dangerous operations that should not be available by default) and for others less so (encoding/decoding primitives, array built-in operations, etc). In any case it's good if the design is consistent across the API. (That's not necessarily the case right now; e.g. `Duktape.act()` is dangerous in sandboxing but available by default to Ecmascript code.)

"Why not both?" is somewhat of a tricky question because for every single decision the impact is small, but there are hundreds of such questions overall and that then begins to matter.

**Comment by svaarala** [comment]
@saghul With Duktape.fin() specifically there's the issue that Ecmascript finalizers are not fully robust (script timeout prevents their execution) so it might be less error prone if only a C API was available. It'd be better to fix the script timeout issue of course :)

**Comment by fatcerberus** [comment]
One advantage to keeping the `Duktape` object is in sandboxing scenarios: Sometimes you want to hide the nonstandard stuff and only allow access to Ecmascript standard functions and methods, and it's very easy to just delete the `Duktape` object as part of that.  With the methods moved to standard object that becomes harder.

One solution to that would be to add a preprocessor define which would leave out the nonstandard functions (or at least not expose them to script).

**Comment by svaarala** [comment]
@fatcerberus True - but note that not all Duktape properties are problematic for sandboxing so there are different issues involved for each of the Duktape object properties. Also some of the standard built-in methods are sandboxing issues right now because of the way internal keys are handled so sticking to standard binding doesn't make the environment sandboxing safe. Current plan is to fix that by restricting access to internal keys to C code only.

**Comment by fatcerberus** [comment]
Sorry, "sandboxing" probably wasn't the right word there.  I just meant the case where someone wants to _only_ expose Ecmascript standard functionality, i.e. No nonstandard features at all.  But that doesn't actually require the Duktape object, a preprocessor define would work just as well.

To be clear, in okay with it if you decide to move the functions.  Just wanted to state what I thought were advantages of the current setup. :)

**Comment by svaarala** [comment]
Ah I see, so basically a compliance option to remove the non-standard bindings.

One concrete case where I'd personally prefer *some* change: `Duktape.enc('jx', value)`. This is much clunkier than, say, `val.toString('jx')` or `val.encode('jx')`. Python has pretty nice and consistent bindings in this respect (`foo.encode('utf-8').encode('hex')` for example) and the current Ecmascript or Duktape ones are not as convenient.

**Comment by fatcerberus** [comment]
Does the standard allow you to extend `.toString()` like that?  That'd be pretty neat if it were possible to call it that way.

**Comment by svaarala** [comment]
That's always a little bit of a grey area in general: there's no way to avoid potential conflicts with future specifications which may add new arguments to existing functions. (That's technically also the case for the `Duktape` object; if ES2030 were to add a "Duktape" global it'd then conflict but in practice this is probably not a very big concern ;-)

For the Object.prototype.toString() case, in ES6 it's not specified to take an argument (http://www.ecma-international.org/ecma-262/6.0/#sec-object.prototype.tostring). In the strict sense the algorithm requires one to ignore any argument (as the steps don't involve the argument). But it's unlikely an argument was passed in practice so extending it this way would be relatively safe I think. The bigger risk is that a future specification version adds an argument to toString() and if that argument cannot be reliably distinguished from an encoding specifier, it'd then be pretty awkward to get out of.

**Comment by fatcerberus** [comment]
The main reason I brought it up is I know there are a few spots in the specification that explicitly disallow any kind of extensions at all, e.g. JSON is very strict about what is allowed.  I wasn't sure if this was one of those cases.

**Comment by saghul** [comment]
@svaarala Thanks for explaining, I understand your point. Keep up the good work :-)

**Comment by svaarala** [comment]
@fatcerberus That's true, for JSON the restrictions are especially strict for the input/output format. Not so much for the API calls or extending the `JSON` object.

In any case extending `Object.prototype.toString()` to take an encoding argument would be more risky than adding a `Object.prototype.encode()` method. It would still be reasonably convenient: `val.encode('jx').encode('base64')` as opposed to `Duktape.enc('base64', Duktape.enc('jx', val))`.

**Comment by fatcerberus** [comment]
I indeed like that `.encode()` call, it makes it very clear what's going on, and the chainability is a nice bonus over `Duktape.enc` :)
